pandas/_libs/tslibs/nattype.pyx

```diff
@@ -103,61 +106,102 @@ cdef class _NaT(datetime):
         if ndim == -1:
             return _nat_scalar_rules[op]
 
-        if ndim == 0:
+        elif util.is_array(other):
```

Comment:
the ```getattr(other, 'ndim', -1)``` is very non-standard (as we almost always make this return 0 by default). can you just use a hasattr check? as the ndim == 0 branch is also never hit as a 0-line scalar hits the is_array branch. (maybe make that clear in a comment). also can pull out line 115 and make it part of the elif

Reply:
yah its weird, but it predates me.  the logic here is kind of weird; id rather take the time to look at it more carefully in a separate pass